modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java

```diff
@@ -267,14 +325,45 @@ public GridMetricManager(GridKernalContext ctx) {
      */
     public MetricRegistry registry(String name) {
         return registries.computeIfAbsent(name, n -> {
-            MetricRegistry mreg = new MetricRegistry(name, log);
+            MetricRegistry mreg = new MetricRegistry(name,
+                mname -> readFromMetastorage(metricName(HITRATE_CFG_PREFIX, mname)),
+                mname -> readFromMetastorage(metricName(HISTOGRAM_CFG_PREFIX, mname)),
+                log);
 
             notifyListeners(mreg, metricRegCreationLsnrs, log);
 
             return mreg;
         });
     }
 
+    /**
+     * Reads value from {@link #metastorage}.
+     *
+     * @param key Key.
+     * @param <T> Key type.
+     * @return Value or {@code null} if not found.
+     */
+    private <T extends Serializable> T readFromMetastorage(String key) {
+        metaLock.readLock().lock();
+
+        try {
+            if (metastorage == null)
+                return null;
+
+            try {
+                return metastorage.read(key);
+            }
+            catch (IgniteCheckedException e) {
```

Comment:
Actually it means that data files are corrupted. Failure process should be invoked in such case. I think it should be done automatically if we pass exception on the next level (but I'm not sure).

Reply:
OK, I changed it to throw `IgniteException` to a higher level.
Anyway, It's a very confusing decision to make "metastore files corrupted" exception checked.